pkg/raft/log.go

```diff
@@ -305,26 +301,35 @@ func (l *raftLog) nextCommittedEnts(allowUnstable bool) (ents []pb.Entry) {
 // hasNextCommittedEnts returns if there is any available entries for execution.
 // This is a fast check without heavy raftLog.slice() in nextCommittedEnts().
 func (l *raftLog) hasNextCommittedEnts(allowUnstable bool) bool {
-	if l.hasNextOrInProgressSnapshot() {
-		// If we have a snapshot to apply, don't also return any committed
-		// entries. Doing so raises questions about what should be applied
-		// first.
-		return false
-	}
-	lo, hi := l.applying+1, l.maxAppliableIndex(allowUnstable)+1 // [lo, hi)
-	return lo < hi
+	return l.applying < l.maxAppliableIndex(allowUnstable)
 }
 
 // maxAppliableIndex returns the maximum committed index that can be applied.
 // If allowUnstable is true, committed entries from the unstable log can be
 // applied; otherwise, only entries known to reside locally on stable storage
 // can be applied.
+//
+// If there is a pending snapshot, the application is paused, and
+// maxAppliableIndex returns l.applying.
 func (l *raftLog) maxAppliableIndex(allowUnstable bool) uint64 {
-	hi := l.committed
-	if !allowUnstable {
-		hi = min(hi, l.unstable.prev.index)
-	}
-	return hi
+	if l.hasNextOrInProgressSnapshot() {
+		// If we have a snapshot to apply, don't return any committed entries. Doing
+		// so raises questions about what should be applied first.
+		//
+		// TODO(pav-kv): the answer to the questions is - the snapshot should be
```

Comment:
How can this happen again? Raft enters the in progress snapshot when it sees the `MsgSnap`. It snap will then come up for application on the next ready cycle, assuming raft accepts the snapshot.
If there are log entries ahead of the snap index, the snapshot will be discarded (might bump the commit idx but not more).
So for a snapshot to come up for application together with entries to apply, we need the following interleaving:

- step MsgSnap
- step MsgApp (let's assume they're accompanied by MsgApp commit idx)
- handle Ready

I guess this is possible in CRDB because to step messages in you don't acquire raftMu? We do hold raftMu across the "step MsgSnap" and "handle Ready" step, but not `repl.mu`.

Reply:
Yeah, this can happen if raft accepts a snapshot and then accepts a log append on top of it. Then we have, say, a snapshot @ `index=100`, and log `(100, 150]`. Today, this code makes sure that we only send the snapshot through `Ready`, and send the entries through the next `Ready`, after the snapshot is applied.